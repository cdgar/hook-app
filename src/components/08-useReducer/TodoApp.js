import React, { useEffect, useReducer } from "react";
import TodoList from "./TodoList";
import todoReducer from "./todoReducer";
import TodoAdd from "./TodoAdd";
import "./styles.css";

const init = () => {
  return JSON.parse(localStorage.getItem("TodoApp")) || [];
};

const TodoApp = () => {
  
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem("TodoApp", JSON.stringify(todos));
  }, [todos]);

  const handleAddTodo = (newTodo) => {
    const actionAdd = { type: "add", payload: newTodo };
    dispatch(actionAdd);
  };

  const handleDeleteTodo = (id) => {
    const actionDel = { type: "delete", payload: id };
    dispatch(actionDel);
  };

  const handleToggleTodo = (id) => {
    const actionDel = { type: "toggle", payload: id };
    dispatch(actionDel);
  };

  return (
    <div>
      <h1>TodoApp ({todos.length})</h1>
      <hr />
      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleToggleTodo={handleToggleTodo}
            handleDeleteTodo={handleDeleteTodo}
          />
        </div>
        <div className="col-5">
          <TodoAdd handleAddTodo={handleAddTodo}/>
        </div>
      </div>
    </div>
  );
};

export default TodoApp;
