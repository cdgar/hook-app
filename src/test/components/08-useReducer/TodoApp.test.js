import React from 'react';
import {shallow, mount } from "enzyme";
import TodoApp from "../../../components/08-useReducer/TodoApp"
import { demoTodos } from '../../fixtures/demoTodos';
import { act } from '@testing-library/react';

describe('TodoApp.js', () => {

    const wrapper = shallow(<TodoApp/>)
    Storage.prototype.setItem = jest.fn(() => {});
    
    test('should show TodoApp', () => {
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should add TodoApp', () => {
        const wrapper = mount(<TodoApp/>);
        act(() => {
            demoTodos.forEach((todo) => wrapper.find('TodoAdd').prop('handleAddTodo')(todo));
        });
        expect(wrapper.find('h1').text().trim()).toBe(`TodoApp (${demoTodos.length})`);
        expect(localStorage.setItem).toHaveBeenCalledTimes(demoTodos.length);    
    })

    test('should delete one todo', () => {
        wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
        wrapper.find('TodoList').prop('handleDeleteTodo')(demoTodos[0].id);
        expect(wrapper.find('h1').text().trim()).toBe(`TodoApp (0)`);
    })
    
    
})