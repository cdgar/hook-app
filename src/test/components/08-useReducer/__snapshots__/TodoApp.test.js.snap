// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoApp.js should show TodoApp 1`] = `
<div>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDeleteTodo={[Function]}
        handleToggleTodo={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
